contracts/pair_stable/src/math.rs

```diff
@@ -73,122 +18,199 @@ pub fn calc_offer_amount(
-            d = calculate_step(&d, leverage, sum_x, &d_product)?;
-            // Equality with the precision of 1
-            if d == d_previous {
-                break;
+            // loop: D_P = D_P * D / (_x * N_COINS)
+            let d_p = pools
+                .iter()
+                .try_fold::<_, _, StdResult<_>>(d, |acc, pool| {
+                    let denominator = pool.atomics().checked_mul(n_coins.into())?;
+                    acc.checked_multiply_ratio(d, Decimal256::new(denominator))
+                })?;
+            let d_prev = d;
+            d = (ann_sum_x + d_p * Decimal256::from_integer(n_coins.u64())) * d
+                / ((ann - Decimal256::one()) * d
+                    + (Decimal256::from_integer(n_coins.u64()) + Decimal256::one()) * d_p);
+            if d >= d_prev {
+                if d - d_prev <= Decimal256::with_precision(Uint64::from(1u8), greatest_precision)?
+                {
+                    return Ok(d);
+                }
+            } else if d < d_prev
+                && d_prev - d <= Decimal256::with_precision(Uint64::from(1u8), greatest_precision)?
+            {
+                return Ok(d);
             }
         }
-        u128::try_from(d).ok()
+
+        Ok(d)
     }
 }
 
 /// ## Description
-/// Helper function used to calculate the D invariant as a last step in the `compute_d` public function.
-///
-/// * **Equation**:
-///
-/// d = (leverage * sum_x + d_product * n_coins) * initial_d / ((leverage - 1) * initial_d + (n_coins + 1) * d_product)
-fn calculate_step(initial_d: &U256, leverage: u64, sum_x: u128, d_product: &U256) -> Option<U256> {
-    let leverage_mul = U256::from(leverage).checked_mul(sum_x.into())? / AMP_PRECISION;
-    let d_p_mul = checked_u8_mul(d_product, N_COINS)?;
-
-    let l_val = leverage_mul.checked_add(d_p_mul)?.checked_mul(*initial_d)?;
-
-    let leverage_sub =
-        initial_d.checked_mul((leverage.checked_sub(AMP_PRECISION)?).into())? / AMP_PRECISION;
-    let n_coins_sum = checked_u8_mul(d_product, N_COINS.checked_add(1)?)?;
-
-    let r_val = leverage_sub.checked_add(n_coins_sum)?;
-
-    l_val.checked_div(r_val)
-}
+/// Computes the new balance of a `to` pool if one makes `from` pool = `new_amount`.
+///
+/// Done by solving quadratic equation iteratively.
+///
+/// `x_1**2 + x_1 * (sum' - (A*n**n - 1) * D / (A * n**n)) = D ** (n + 1) / (n ** (2 * n) * prod' * A)`
+///
+/// `x_1**2 + b*x_1 = c`
+///
+/// `x_1 = (x_1**2 + c) / (2*x_1 + b)`
+pub(crate) fn calc_y(
+    from_asset: &DecimalAsset,
+    to: &AssetInfo,
+    new_amount: Decimal256,
+    pools: &[DecimalAsset],
+    amp: Uint64,
+    target_precision: u8,
+) -> StdResult<Uint128> {
+    if to.equal(&from_asset.info) {
+        return Err(StdError::generic_err(
+            "The offer asset and ask asset cannot be the same.",
+        ));
+    }
+    if from_asset.amount.eq(&new_amount) {
```

Comment:
Why do we have this check? This is only possible when from_pool is empty but we cover this case in check_swap_parameters() function.

Reply:
True, removed redundant check, thanks